src/group_theory/submonoid/operations.lean

```diff
@@ -896,3 +896,146 @@ instance [mul_action M' α] [has_faithful_scalar M' α] (S : submonoid M') :
 end submonoid
 
 end actions
+
+-- For the pointwise inverse for submonoids of groups,
+-- please refer to `group_theory.submonoid.pointwise`.
+section pointwise_inverse
+
+namespace submonoid
+
+@[to_additive] noncomputable
+instance {M : Type*} [monoid M] : group (is_unit.submonoid M) :=
+{ inv := λ x, ⟨_, (x.prop.unit⁻¹).is_unit⟩,
+  mul_left_inv := λ x, subtype.eq x.prop.unit.inv_val,
+ ..(show monoid (is_unit.submonoid M), by apply_instance) }
+
+
+@[to_additive] noncomputable
+instance {M : Type*} [comm_monoid M] : comm_group (is_unit.submonoid M) :=
+{ mul_comm := λ a b, mul_comm a b,
+ ..(show group (is_unit.submonoid M), by apply_instance) }
+
+section monoid
+
+variables {M' : Type*} [monoid M'] (S' : submonoid M')
+
+/-- `S'.left_inv` is the submonoid containing all the left inverses of `S'`. -/
+@[to_additive "`S'.left_neg` is the additive submonoid containing all the left additive inverses
+of `S'`."]
+def left_inv : submonoid M' :=
+{ carrier := { x : M' | ∃ y : S', x * y = 1 },
+  one_mem' := ⟨1, mul_one 1⟩,
+  mul_mem' := λ a b ⟨a', ha⟩ ⟨b', hb⟩,
+    ⟨b' * a', by rw [coe_mul, ← mul_assoc, mul_assoc a, hb, mul_one, ha]⟩ }
+
+@[to_additive]
+lemma left_inv_left_inv_le : S'.left_inv.left_inv ≤ S' :=
+begin
+  rintros x ⟨⟨y, z, h₁⟩, h₂ : x * y = 1⟩,
+  convert z.prop,
+  rw [← mul_one x, ← h₁, ← mul_assoc, h₂, one_mul],
+end
+
+@[to_additive]
+lemma unit_mem_left_inv (x : units M') (hx : (x : M') ∈ S') : ((x⁻¹ : _) : M') ∈ S'.left_inv :=
+⟨⟨x, hx⟩, x.inv_val⟩
+
+attribute [to_additive] is_unit.submonoid is_unit.unit
+
+@[to_additive]
+lemma left_inv_left_inv_eq (hS' : S' ≤ is_unit.submonoid M') : S'.left_inv.left_inv = S' :=
+begin
+  refine le_antisymm S'.left_inv_left_inv_le _,
+  intros x hx,
+  have : x = ((hS' hx).unit⁻¹⁻¹ : units M') := by { rw [inv_inv (hS' hx).unit], refl },
+  rw this,
+  exact S'.left_inv.unit_mem_left_inv _ (S'.unit_mem_left_inv _ hx)
+end
+
+/-- The function from `S'.left_inv` to `S'` sending an element to its right inverse in `S'`.
+This is a `monoid_hom` when `M'` is commutative. -/
+@[to_additive "The function from `S'.left_add` to `S'` sending an element to its right additive
+inverse in `S'`. This is an `add_monoid_hom` when `M'` is commutative."]
+noncomputable
+def from_left_inv : S'.left_inv → S' := λ x, x.prop.some
```

Comment:
If you make this
```suggestion
def from_left_inv : S'.left_inv → S'ᵐᵒᵖ := λ x, x.prop.some
```
Then it is always a monoid_hom

Reply:
In order to have such a `monoid_hom`, I would probably need some more hypotheses to guarantee that the inverse is unique, since the inverse is obtained by choice.